d1985r0.md

```diff
@@ -19,18 +21,27 @@ type traits.
 
 # Change Log
 
+## R1 -> R2
+
+Having found overwhelming support for the feature in EWGI, and a concern
+about contra- and co-variance of the `template auto` parameters, we include
+discussion of these topics in the paper, together with the decision and
+comparison tables that underpin the decision.
+
+The decision was to leave the co/contravariance as is.
+
 ## R0 -> R1
 
 - Greatly expanded the number of examples based on feedback from the BSI panel.
 - Clarified that we are proposing eager checking
 
 # Motivation
 
-Imagine trying to write a metafunction for apply. While `apply` is very
+Imagine trying to write the `apply` metafunction. While `apply` is very
 simple, a metafunction like `bind` or `curry` runs into the same problems;
 for demonstration, `apply` is clearest.
```

Comment:
fixed:
- simplicity
- symmetry
- metafunction such a
- using value = Arg
- template-template
- moved curry thing
- standardized on typename

Deferring discussion of co/contravariance to when it happens.

Reply:
Re eager/late checking: I've had a nice chat with Michael Park this morning.

I think the way we should phrase it is "it's late checking, but the symbol kind is a special kind that doesn't really admit any operations other than passing it as an argument". That is, *maybe* we could allow "casts" with `typename` and `template`, but I haven't thought about that yet. 

Re co/contravariance: would you have time to get on a call or something? I think this would be better to solve with an interactive session.

Reply:
Re: co/contravariance, sure, I'd be up for a call, but I'll get my thoughts down here as text first:

Permitting `template <typename...> typename` to bind to non-variadic templates was a mistake.  I'd like to see us provide another mechanism, and deprecate that binding.  The mental model should be (and currently is, kinda) that the declaration of a template template parameter defines an interface that can be used from within the template definition.  For instance, this is already valid:
```c++
template <template <typename T> class TT, typename T>
auto foo()
{
    return TT<T>::x;
}

template <typename T, typename U = std::decay_t<T>>
struct S
{
    static constexpr auto x = sizeof(U);
};

std::size_t f()
{
    return foo<S, int[10]>;	// returns sizeof(int*)
}
```

What we need is a way to specify _permissive_ (or _less constrained_) template parameters.  I think we can get there by expanding on our notion of template template parameters:

* `typename T`: type parameter, as before
* `auto V`: non-type parameter, as before
* `template <typename T> typename TT`: template parameter; binds to type templates accepting a single type argument
* `template <typename... Ts> typename TT`: binds only to type templates accepting an unbounded number of type parameters (deprecate existing "binds to templates having any combination of type parameters")
* `template typename TT`: unrestricted template parameter; binds to any type template
* `template <__ X> typename TT`: binds to any type template accepting a single template parameter of uspecified kind
* `template <template auto X> typename TT`: binds to any template accepting a single template parameter of any kind

This is what `apply` would look like:
```c++
// Don't require F to accept any number of template arguments
template <template typename F, template auto... Args>
using apply = F<Args...>;
```

...and here's `fwd`:
```c++
template <template <__ X> typename F, template auto Arg>
struct fwd {
  using type = F<Arg>; // ok for phase 1, may be invalid in phase 2
};
```

Reply:
I'm liking this a lot, actually, the more I read it. Spellings might be an issue, still, but ...